File: benchmarks/go/churn_bench.go
// Benchmark: alloc/drop churn — constantly create and discard objects.
// Measures GC throughput under sustained allocation pressure.
//
// Usage: go run common.go churn_bench.go [N]
package main

import (
	"runtime"
	"time"
)

type ChurnNode struct {
	Value int
}

func main() {
	n := mustGetN()

	runtime.GC()

	start := time.Now()

	// Keep a window of live objects, constantly replacing oldest
	windowSize := 1000
	window := make([]*ChurnNode, windowSize)
	totalAllocs := 0

	for i := 0; i < n; i++ {
		window[i%windowSize] = &ChurnNode{Value: i}
		totalAllocs++
	}

	// Prevent dead-code elimination
	sum := 0
	for _, obj := range window {
		if obj != nil {
			sum += obj.Value
		}
	}
	_ = sum

	elapsed := time.Since(start)
	peakHeap, heapInUse, numGC, totalPause, maxPause := gcStats()

	printResult(BenchResult{
		Benchmark:    "churn",
		Language:     "go",
		Duration:     elapsed,
		PeakHeapMB:   peakHeap,
		HeapInUseMB:  heapInUse,
		NumGC:        numGC,
		TotalGCPause: totalPause,
		MaxGCPause:   maxPause,
		Throughput:   float64(totalAllocs) / elapsed.Seconds(),
	})
}

File: benchmarks/rust/src/bin/churn_bench.rs
// Benchmark: alloc/drop churn — constantly create and discard objects.
//
// Usage: cargo run --release --bin churn_bench -- [N] [--strategy=...]

use ferris_gc::{Finalize, Gc, Trace, LOCAL_GC};
use ferris_gc_benchmarks::*;

struct ChurnNode {
    value: i32,
}

impl Trace for ChurnNode {
    fn is_root(&self) -> bool { false }
    fn reset_root(&self) {}
    fn trace(&self) {}
    fn reset(&self) {}
    fn is_traceable(&self) -> bool { false }
}

impl Finalize for ChurnNode {
    fn finalize(&self) {}
}

fn main() {
    let _cleanup = ferris_gc::ApplicationCleanup;
    let n = get_n();
    let strategy = configure_local_strategy(&get_strategy());
    install_local_gc_monitor();

    LOCAL_GC.with(|gc| unsafe { gc.borrow()._collect() });

    let window_size = 1000;

    let (_, elapsed) = timed(|| {
        let mut window: Vec<Option<Gc<ChurnNode>>> = (0..window_size).map(|_| None).collect();

        for i in 0..n {
            window[i % window_size] = Some(Gc::new(ChurnNode { value: i as i32 }));
        }

        // Prevent optimization
        let sum: i64 = window
            .iter()
            .filter_map(|opt| opt.as_ref())
            .map(|gc| (**gc).value as i64)
            .sum();
        std::hint::black_box(sum);
    });

    let stats = local_gc_stats();
    make_result("churn", &strategy, n, elapsed, &stats).print();
}
